Roointerface, BATCalculator

fredRos opened the pull request:
refactor and reactivate examples. Fixes #100

oschulz commented:
Can't merge due to conflicts.

fredRos commented:
It turns out there are more issues

thread safety
-----------------

Both `BCRooInterface` and `BATCalculator` fail with a segfault with multiple threads. The serial version with `OMP_NUM_THREADS=1` still works. At the moment I'm fine with stating that this interface is not thread safe for 1.0.

 @oschulz I guess we can just set the number of threads to 1 by hand in the example, so the tests can all run with multiple threads except this one. Is that ok with you?

root import
--------------
It works with root 5.34.19 but with root 6.05.02 this is what happens
```
================================================================
   BAT 0.9.4.1: examples/advanced/rooInterface/test-suite.log
================================================================

# TOTAL: 5
# PASS:  4
# SKIP:  0
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0

.. contents:: :depth: 2

FAIL: runBATCalculator
======================


Processing ./runBATCalculator.C...

ESC[1mRooFit v3.60 -- Developed by Wouter Verkerke and David KirkbyESC[0m 
                Copyright (C) 2000-2013 NIKHEF, University of California & Stanford University
                All rights reserved, please read http://roofit.sourceforge.net/license.txt

In file included from input_line_9:1:
/home/beaujean/workspace/bat/examples/advanced/rooInterface/runBATCalculator.C:120:9: error: unknown type name 'BATCalculator'
        BATCalculator batcalc(data, modelconfig);
```
I suspect that these incantations in `BATCalculator.h`
```c++
ClassDef(BATCalculator, 1) // BATCalculator class
```
and `BATCalculator.cxx`
```c++
ClassImp(RooStats::BATCalculator)
```
only work for root 5. @oschulz , what do you think? Is there are good way to fix that without days of work?

oschulz commented:
> @oschulz I guess we can just set the number of threads to 1 by hand in the example, so the tests can all run with multiple threads except this one. Is that ok with you?

Sure - better to have these examples in the test with `OMP_NUM_THREADS=1`, than leaving them disabled.

oschulz commented:
> ClassImp(RooStats::BATCalculator)

AFAIK this should not be needed - ClassImp is mostly used for the ROOT-style API documentation (API-doc comments in .cpp instead of .h). However, it should still work fine with ROOT-6.

> ClassDef(BATCalculator, 1)

If we're not planning to write BATCalculator objects to a TFile, we can set the class version no. to 0. Maybe that helps.

fredRos commented:
So now the test run with a single thread and `runBATCalculator` is marked as known to fail for root 6. I was looking at the build process and polished a few things here and there but nothing major. Please test and merge now

oschulz commented:
> runBATCalculator is marked as known to fail for root 6

Hm, this doesn't seem to work for me. When I run `make distcheck` (Linux, GCC, ROOT-6), it fails:

```
make[6]: *** [WS_Poisson_withSystematics.root] Error 1
root [0]
Processing ../../../../examples/advanced/rooInterface/PrepareWorkspace_GaussOverFlat.C...
XFAIL: runBATCalculator.C
```

oschulz commented on examples/advanced/rooInterface/Makefile.am:
```diff
@@ -13,13 +21,17 @@ input_workspaces = \
+
+# only works with root 5
+if WITH_CLING
+XFAIL_TESTS = runBATCalculator.C
```
We should just leave it out, with ROOT-6, instead of adding it to `XFAIL_TESTS` - the latter carries the intention that the test *should* fail. Meaning, when it will pass, some day, because of other fixes, the build test as a whole would suddenly fail.

fredRos replied:
> when it will pass, some day, because of other fixes

Ok, it won't accidentally work I guess :smile: 
I wonder why it fails, perhaps your test driver can't deal with XFAIL...
In any case, I will just deactivate the test for root6. Assuming that you still test  with root5, at least we can be sure it is tested before release even though users of root6 can't run the example. Oh well, let's wait if anyone complains.

oschulz commented:
Passes on OS-X (ROOT-5 and -6, non-parallel BAT), but fails on Linux (ROOT-5 and -6, parallel BAT):

```
root -l -x -q ../../../../examples/advanced/rooInterface/PrepareWorkspace_Poisson_withSystematics.C
[...]
make[6]: *** [WS_GaussOverFlat.root] Error 1
make[6]: *** Waiting for unfinished jobs....
make[6]: *** [WS_GaussOverFlat_withSystematics.root] Error 1
make[6]: *** [WS_Poisson.root] Error 1
```

fredRos commented:
Sorry I'm unable to reproduce that problem. Could it be some issue with a filesystem on the machine where you test? In any case, I wanted to consolidate the changes, and I guess this one takes more effort than #128, so I rebased this one onto #128. I now added a few more fixes  from today.

oschulz commented:
> Could it be some issue with a filesystem on the machine where you test?

No, I run the tests in /tmp (resp. `$TMPDIR`), which is is ordinary local storage.

fredRos commented:
No I was able to recreate your test setup and experience the same behavior,  the tail of the log says it all

```
make[6]: Entering directory `/tmp/beaujean-test-bat-1o5PL9/src/bat/BAT-0.9.4.1/_build/examples/advanced/rooInterf\
ace'
/.hb/raidg01/sw/root/linux-ubuntu-14.04-x86_64/5.34.34/bin/root -l -x -q ../../../../examples/advanced/rooInterfa\
ce/PrepareWorkspace_GaussOverFlat.C
/.hb/raidg01/sw/root/linux-ubuntu-14.04-x86_64/5.34.34/bin/root -l -x -q ../../../../examples/advanced/rooInterfa\
ce/PrepareWorkspace_GaussOverFlat_withSystematics.C
/.hb/raidg01/sw/root/linux-ubuntu-14.04-x86_64/5.34.34/bin/root -l -x -q ../../../../examples/advanced/rooInterfa\
ce/PrepareWorkspace_Poisson.C
/.hb/raidg01/sw/root/linux-ubuntu-14.04-x86_64/5.34.34/bin/root -l -x -q ../../../../examples/advanced/rooInterfa\
ce/PrepareWorkspace_Poisson_withSystematics.C
/.hb/raidg01/sw/root/linux-ubuntu-14.04-x86_64/5.34.34/bin/root: can't figure out DISPLAY, set it manually
In case you run a remote ssh session, restart your ssh session with:
=========>  ssh -Y
make[6]: *** [WS_GaussOverFlat.root] Error 1
make[6]: *** Waiting for unfinished jobs....
/.hb/raidg01/sw/root/linux-ubuntu-14.04-x86_64/5.34.34/bin/root: can't figure out DISPLAY, set it manually
In case you run a remote ssh session, restart your ssh session with:
=========>  ssh -Y
make[6]: *** [WS_GaussOverFlat_withSystematics.root] Error 1
/.hb/raidg01/sw/root/linux-ubuntu-14.04-x86_64/5.34.34/bin/root: can't figure out DISPLAY, set it manually
In case you run a remote ssh session, restart your ssh session with:
=========>  ssh -Y
make[6]: *** [WS_Poisson.root] Error 1
/.hb/raidg01/sw/root/linux-ubuntu-14.04-x86_64/5.34.34/bin/root: can't figure out DISPLAY, set it manually
In case you run a remote ssh session, restart your ssh session with:
=========>  ssh -Y
make[6]: *** [WS_Poisson_withSystematics.root] Error 1
PASS: runBATCalculator.C
make[6]: Leaving directory `/tmp/beaujean-test-bat-1o5PL9/src/bat/BAT-0.9.4.1/_build/examples/advanced/rooInterfa\
ce'
make[5]: Leaving directory `/tmp/beaujean-test-bat-1o5PL9/src/bat/BAT-0.9.4.1/_build/examples/advanced/rooInterfa\
ce'                                                                                                               
make[5]: *** [check-TESTS] Error 2
make[4]: Leaving directory `/tmp/beaujean-test-bat-1o5PL9/src/bat/BAT-0.9.4.1/_build/examples/advanced/rooInterfa\
ce'
make[4]: *** [check-am] Error 2
make[3]: Leaving directory `/tmp/beaujean-test-bat-1o5PL9/src/bat/BAT-0.9.4.1/_build/examples/advanced'
make[3]: *** [check-recursive] Error 1
make[2]: Leaving directory `/tmp/beaujean-test-bat-1o5PL9/src/bat/BAT-0.9.4.1/_build/examples'
make[2]: *** [check-recursive] Error 1
make[1]: *** [check-recursive] Error 1
make[1]: Leaving directory `/tmp/beaujean-test-bat-1o5PL9/src/bat/BAT-0.9.4.1/_build'
make: *** [distcheck] Error 1
FAIL
ERROR: Build, install or test(s) failed, test area "/tmp/beaujean-test-bat-1o5PL9".
```

So apparently this is not due to the code but to the ssh connection used in the test build. So I think we can then merge this pull request soon.
